src/app/services/covid-data-api.service.ts

```diff
@@ -1,58 +1,97 @@
 import { Injectable, Inject, InjectionToken } from '@angular/core';
 import { CovidDataPoint } from '../models/CovidDataPoint';
+import { Country } from '../models/Country';
 
 import { Observable, EMPTY } from 'rxjs';
 import { catchError } from 'rxjs/operators';
 
-export interface Country {
-  // TODO: Move this somewhere where it makes more sense (maybe the country picker component)
-  id: number;
-  name: string;
-  iso2: string;
-  iso3: string;
-  continent: string;
-  lat: string;
-  long: string;
-}
-
-export interface CovidDataApiSubService {
+interface CovidDataApiServiceInterface {
   getLatestGlobalData(): Observable<CovidDataPoint>;
   getLatestCountryData(country: Country): Observable<CovidDataPoint>;
 }
 
+export interface CovidDataApiSubService extends CovidDataApiServiceInterface {}
+
 export const COVID_DATA_API_SUB_SERVICE = new InjectionToken<
   CovidDataApiSubService
 >('CovidDataApiSubService');
 
 @Injectable({
   providedIn: 'root',
 })
-export class CovidDataApiService {
+export class CovidDataApiService implements CovidDataApiServiceInterface {
   constructor(
     // Inspired by (https://stackoverflow.com/a/35916788)
     // Each api service must have its own provider registered in app.module for the InjectionToken
     @Inject(COVID_DATA_API_SUB_SERVICE)
-    private apiService: CovidDataApiSubService[]
+    private apiServices: CovidDataApiSubService[]
   ) {}
 
+  /**
+   * Cycles through each api service, calling the method provided, till one returns a value
+   * If all lead to errors, returns an empty observable
+   * @param {Observable<T>} method The method to be called
+   */
+  private getResultFromOneApiFor<T>(
+    method: (...args: any[]) => Observable<T>, // reference to a method with the same name being called
+    ...args: any[]
+  ): Observable<T> {
+    // if the method is not a class method of this class throw an error
+    if (
+      !Object.getOwnPropertyNames(CovidDataApiService.prototype).includes(
+        method.name
+      )
+    ) {
+      throw new Error(
+        `Method: ${method.name}() not found in Object: ${this.constructor.name}`
+      );
+    }
```

Comment:
Not a big fan of this solution. If you want to use the same function for different methods, you may wish to try using a command pattern, i.e. :
```
interface ApiFetchCommand {
  call(): Observable<CovidDataPoint>;
}
class GetLatestCountryDataCommand implements ApiFetchCommand {
  constructor(private api: CovidDataApiSubservice, private country: Country) {}

  call(): Observable<CovidDataPoint> {
    return this.api.getLatestCountryData(this.country);
  }
}
```
and then you can simply wrap each subservice inside the command before processing it.

Alternatively, and perhaps more simply, you can combine both functions into 1 taking region as a parameter, with 'global' returning world data.

Reply:
I wasn't that big of a fan of it either at first, but wanted to find a way to be able to call the same method on each of the sub services. 
But also wanted it to be generic so that if we add more fetch methods, then it takes care of it itself without much setup code for the new ones (as long as the sub service implements the `CovidDataApiServiceInterface` and the new fetch methods are part of it).
This probably makes the result a bit cryptic to understand though.

This command pattern seems interesting. Although we will need to create new command objects for any new commands.

Reply:
What I don't really understand is why we need more than 1 fetch method.
Here you distinguish between global and country statistics, but where is the information which of these to use coming from? If we decide to use a region service we can inject it here and then the function can decide what to do internally (if global, call global endpoint, else call country endpoint).

The only other thing that comes to mind is that we may want to grab historical data from the api. That would probably be a separate class though, in line with single responsibility principle.

Reply:
> What I don't really understand is why we need more than 1 fetch method.
> Here you distinguish between global and country statistics, but where is the information which of these to use coming from? If we decide to use a region service we can inject it here and then the function can decide what to do internally (if global, call global endpoint, else call country endpoint).
> 
> The only other thing that comes to mind is that we may want to grab historical data from the api. That would probably be a separate class though, in line with single responsibility principle.

I guess it depends if we define the responsibility as 'get data from the api' or 'get latest data from the api'

Reply:
The former seems to result in an extra layer of complication using the command pattern, so it may be a symptom that we're trying to do too many things at once.
Anyway, I'm not going to complain too much, if you're happy using the command pattern for the api calls, then we can certainly go with that!